File: backend/internal/modules/bom/model/bom.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// BOMView BOM视图模型
type BOMView struct {
	ID             uint           `gorm:"primaryKey" json:"id"`
	Name           string         `gorm:"size:200;not null" json:"name"`
	RootMaterialID uint           `gorm:"not null;uniqueIndex:uk_root_material" json:"root_material_id"`
	RootVersion    string         `gorm:"size:2" json:"root_version"`
	IsExact        bool           `gorm:"not null;default:false" json:"is_exact"`
	Status         string         `gorm:"size:20;default:DRAFT;index" json:"status"`
	Description    string         `json:"description"`
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	CreatedBy      uint           `gorm:"not null" json:"created_by"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	RootMaterial *Material    `gorm:"foreignKey:RootMaterialID" json:"root_material,omitempty"`
	Items        []BOMItem    `gorm:"foreignKey:BOMViewID" json:"items,omitempty"`
}

// BOMItem BOM项模型
type BOMItem struct {
	ID          uint      `gorm:"primaryKey" json:"id"`
	BOMViewID   uint      `gorm:"not null;index" json:"bom_view_id"`
	ParentID    *uint     `gorm:"index" json:"parent_id"`
	MaterialID  uint      `gorm:"not null;index" json:"material_id"`
	Version     string    `gorm:"size:2" json:"version"`
	Quantity    float64   `gorm:"type:decimal(10,4);not null;default:1" json:"quantity"`
	SortOrder   int       `gorm:"default:0" json:"sort_order"`
	Level       int       `gorm:"default:1" json:"level"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`

	// 关联
	Material *Material  `gorm:"foreignKey:MaterialID" json:"material,omitempty"`
	Children []BOMItem  `gorm:"foreignKey:ParentID" json:"children,omitempty"`
}

// Material 简化的物料模型（用于关联查询）
type Material struct {
	ID       uint   `gorm:"primaryKey" json:"id"`
	ItemID   string `gorm:"size:50;not null" json:"item_id"`
	ItemName string `gorm:"size:200;not null" json:"item_name"`
	ItemType string `gorm:"size:50" json:"item_type"`
	Version  string `gorm:"size:2" json:"version"`
	Unit     string `gorm:"size:20" json:"unit"`
}

// BOMTreeNode BOM树节点（用于前端展示）
type BOMTreeNode struct {
	ID         uint          `json:"id"`
	MaterialID uint          `json:"material_id"`
	ItemID     string        `json:"item_id"`
	ItemName   string        `json:"item_name"`
	Version    string        `json:"version"`
	Quantity   float64       `json:"quantity"`
	SortOrder  int           `json:"sort_order"`
	Level      int           `json:"level"`
	Unit       string        `json:"unit"`
	Children   []BOMTreeNode `json:"children,omitempty"`
}

// CreateBOMViewRequest 创建BOM视图请求
type CreateBOMViewRequest struct {
	Name           string  `json:"name" binding:"required,max=200"`
	RootMaterialID uint    `json:"root_material_id" binding:"required"`
	RootVersion    string  `json:"root_version"`
	IsExact        bool    `json:"is_exact"`
	Description    string  `json:"description"`
}

// UpdateBOMViewRequest 更新BOM视图请求
type UpdateBOMViewRequest struct {
	Name        string `json:"name" binding:"max=200"`
	Status      string `json:"status" binding:"omitempty,oneof=DRAFT REVIEWING RELEASED REJECTED OBSOLETE"`
	Description string `json:"description"`
}

// AddBOMItemRequest 添加BOM项请求
type AddBOMItemRequest struct {
	ParentID   *uint   `json:"parent_id"`
	MaterialID uint    `json:"material_id" binding:"required"`
	Version    string  `json:"version"`
	Quantity   float64 `json:"quantity" binding:"required,gt=0"`
	SortOrder  int     `json:"sort_order"`
}

// UpdateBOMItemRequest 更新BOM项请求
type UpdateBOMItemRequest struct {
	Quantity  float64 `json:"quantity" binding:"required,gt=0"`
	SortOrder int     `json:"sort_order"`
}

// ConvertBOMRequest 转换BOM类型请求
type ConvertBOMRequest struct {
	ToExact bool `json:"to_exact"`
}

// BOMListQuery BOM列表查询参数
type BOMListQuery struct {
	Page     int    `form:"page"`
	PageSize int    `form:"page_size"`
	Name     string `form:"name"`
	Status   string `form:"status"`
	IsExact  *bool  `form:"is_exact"`
}

// BOMViewResponse BOM视图响应
type BOMViewResponse struct {
	BOMView
	ItemCount int `json:"item_count"`
}

// ImportBOMResult 导入BOM结果
type ImportBOMResult struct {
	SuccessCount int      `json:"success_count"`
	FailCount    int      `json:"fail_count"`
	Errors       []string `json:"errors,omitempty"`
}

// TableName 指定表名
func (BOMView) TableName() string {
	return "bom_views"
}

func (BOMItem) TableName() string {
	return "bom_items"
}

func (Material) TableName() string {
	return "materials"
}

File: frontend/src/api/bom.ts
import { request } from '@/utils/request'
import type { ApiResponse, PageData } from '@/types/api'

// BOM视图类型
export interface BOMView {
  id: number
  name: string
  root_material_id: number
  root_version: string
  is_exact: boolean
  status: string
  description: string
  created_at: string
  updated_at: string
  created_by: number
  root_material?: Material
  item_count?: number
}

// BOM项类型
export interface BOMItem {
  id: number
  bom_view_id: number
  parent_id: number | null
  material_id: number
  version: string
  quantity: number
  sort_order: number
  level: number
  material?: Material
  children?: BOMItem[]
}

// BOM树节点
export interface BOMTreeNode {
  id: number
  material_id: number
  item_id: string
  item_name: string
  version: string
  quantity: number
  sort_order: number
  level: number
  unit: string
  children?: BOMTreeNode[]
}

// 物料简要信息
export interface Material {
  id: number
  item_id: string
  item_name: string
  item_type: string
  version: string
  unit: string
}

// BOM列表查询参数
export interface BOMListParams {
  page?: number
  page_size?: number
  name?: string
  status?: string
  is_exact?: boolean
}

// 创建BOM视图请求
export interface CreateBOMViewRequest {
  name: string
  root_material_id: number
  root_version?: string
  is_exact?: boolean
  description?: string
}

// 更新BOM视图请求
export interface UpdateBOMViewRequest {
  name?: string
  status?: string
  description?: string
}

// 添加BOM项请求
export interface AddBOMItemRequest {
  parent_id?: number
  material_id: number
  version?: string
  quantity: number
  sort_order?: number
}

// 更新BOM项请求
export interface UpdateBOMItemRequest {
  quantity: number
  sort_order?: number
}

// 转换BOM类型请求
export interface ConvertBOMRequest {
  to_exact: boolean
}

// 获取BOM视图列表
export function getBOMList(params: BOMListParams): Promise<ApiResponse<PageData<BOMView>>> {
  return request.get('/bom/boms', { params })
}

// 创建BOM视图
export function createBOMView(data: CreateBOMViewRequest): Promise<ApiResponse<BOMView>> {
  return request.post('/bom/boms', data)
}

// 获取BOM视图详情
export function getBOMDetail(id: number): Promise<ApiResponse<BOMView>> {
  return request.get(`/bom/boms/${id}`)
}

// 更新BOM视图
export function updateBOMView(id: number, data: UpdateBOMViewRequest): Promise<ApiResponse<BOMView>> {
  return request.put(`/bom/boms/${id}`, data)
}

// 删除BOM视图
export function deleteBOMView(id: number): Promise<ApiResponse<null>> {
  return request.delete(`/bom/boms/${id}`)
}

// 获取BOM树形结构
export function getBOMTree(id: number): Promise<ApiResponse<BOMTreeNode>> {
  return request.get(`/bom/boms/${id}/tree`)
}

// 添加BOM项
export function addBOMItem(bomId: number, data: AddBOMItemRequest): Promise<ApiResponse<BOMItem>> {
  return request.post(`/bom/boms/${bomId}/items`, data)
}

// 更新BOM项
export function updateBOMItem(
  bomId: number,
  itemId: number,
  data: UpdateBOMItemRequest
): Promise<ApiResponse<BOMItem>> {
  return request.put(`/bom/boms/${bomId}/items/${itemId}`, data)
}

// 删除BOM项
export function deleteBOMItem(bomId: number, itemId: number): Promise<ApiResponse<null>> {
  return request.delete(`/bom/boms/${bomId}/items/${itemId}`)
}

// 转换BOM类型
export function convertBOMType(id: number, data: ConvertBOMRequest): Promise<ApiResponse<null>> {
  return request.post(`/bom/boms/${id}/convert`, data)
}

// 导出BOM
export function exportBOM(id: number): Promise<ApiResponse<{ download_url: string }>> {
  return request.get(`/bom/boms/${id}/export`)
}

// 导入BOM
export function importBOM(file: File): Promise<ApiResponse<{ success_count: number; fail_count: number }>> {
  const formData = new FormData()
  formData.append('file', file)
  return request.post('/bom/import', formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  })
}
